Fixed OS tag in ZipArchive header

ianhays opened the pull request:
When writing, the OS tag in the header of a ZipArchive CentralDirectoryHeader was being set to 0 (Windows) regardless of platform. When reading, the os tag was being ignored. I added logic to set/get this tag properly and also modified the ArchiveEntry Name getter to parse the path based on the source OS instead of on the running OS.

This will break anyone that was uzing the Name property of ZipArchiveEntries that had paths containing the backslash character within a ZipArchive that was compressed on a Unix machine but decompressed on a Windows machine.

resolves #1904. Related to #4991 

@stephentoub @justinvp @mellinoe

justinvp commented on src/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs:
```diff
@@ -110,7 +113,13 @@ internal ZipArchiveEntry(ZipArchive archive, String entryName)
 
             _diskNumberStart = 0;
-
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux) || RuntimeInformation.IsOSPlatform(OSPlatform.OSX))
```
This library is currently compiled separately for Unix and Windows, so technically Unix vs. Windows could be determined at compile time instead of taking a dependency on `System.Runtime.InteropServices.RuntimeInformation`. But I'm guessing the use of `RuntimeInformation` is still preferred?

ianhays replied:
I thought about splitting the file so we would have a ZipArchiveEntry.Unix and ZipArchiveEntry.Windows but I went with this route since it was more intuitive. Thinking over it more, you're right that it is probably better to not depend on RuntimeInformation if we can avoid i (and we can). I'll split the file.

justinvp commented on src/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs:
```diff
@@ -250,8 +259,41 @@ public String Name
+                // equal to this OS. Otherwise we resort to manual parsing.
+                string path = FullName;
+                Debug.Assert(path != null);
+                if (path.IndexOfAny(Path.GetInvalidPathChars()) >= 0)  // Validation checking is done based on current OS, not source OS.
```
`Path.GetInvalidPathChars()` always allocates and return a new array. Should `PathInternal.InvalidPathChars` under src/Common be used instead, to avoid this?

ianhays replied:
Thanks. I forgot about adding PathInternal. I should be using PathInternal.HasIllegalCharacters(path).

justinvp commented on src/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs:
```diff
@@ -23,8 +23,9 @@ public class ZipArchiveEntry
         private readonly Boolean _originallyInArchive;
         private readonly Int32 _diskNumberStart;
+        private ZipVersionMadeByPlatform _versionMadeByPlatform;
+        private byte _versionMadeBySpecification;
```
Can these be readonly?

ianhays replied:
for sure.

ianhays commented:
@dotnet-bot test this please

stephentoub commented on src/System.IO.Compression/src/System/IO/Compression/ZipVersion.cs:
```diff
@@ -0,0 +1,24 @@
+    {
+        Windows = 0,
+        Unix = 3,
+        Undefined = 255
```
Is Undefined used anywhere anymore?

ianhays replied:
No it is not since I split the files. Removing.

stephentoub commented on src/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs:
```diff
@@ -65,6 +65,8 @@ internal ZipArchiveEntry(ZipArchive archive, ZipCentralDirectoryFileHeader cd)
             _originallyInArchive = true;
 
             _diskNumberStart = cd.DiskNumberStart;
+            _versionMadeByPlatform = (ZipVersionMadeByPlatform)cd.VersionMadeByCompatibility;
```
What happens if this ZipArchiveEntry is then saved out again?  What value ends up getting written out, and is that the correct value?

ianhays replied:
Good catch! The value written out will be the value that we initialized it to in the constructor. So if we have a zip with OS version "13" and read it on a Windows system and then write it again the written zip will have the OS set to "13".

The OS tag is <i>supposed</i> to be set to the system that the zip was written on, so this behavior is incorrect. It's possible that it would be okay in the scenario where the zip is opened and rewritten without modification, but i could see that causing more issues than it would circumvent.

I'm going to modify the write code to always write the CurrentZipPlatform.

stephentoub commented on src/System.IO.Compression/tests/ZipArchive/zip_InvalidParametersAndStrangeFiles.cs:
```diff
@@ -185,12 +185,11 @@ public static void InvalidConstructors()
             ZipTest.IsZipSameAsDir(await StreamHelpers.CreateTempCopyStream(
                  ZipTest.strange("dataDescriptor.zip")), ZipTest.zfolder("normalWithoutBinary"), ZipArchiveMode.Update, true, false);
-	    }
+        }
```
Do we have any tests that create appropriate archives on one OS and validate them on another?  If we don't, I expect this would involve creating a few additional .zip files and adding them to the test assets we pull down as part of test runs.

ianhays replied:
We have some cross-compat zip tests but they do not check the validity of the header tags, only the files/folders within the zip. I have an issue (https://github.com/dotnet/corefx/issues/4992) up for creating more cross-compat tests that cover scenarios that are valid on one system but invalid on the other. 

As part of that addition I'd like to increase the robustness of our ZipArchive validation (e.g. check the header tags for the central directory header) on the Zips that we're writing. It'll be a bit more in depth since we don't expose any API to get the OS version tag, so I'd planned on putting it in its own PR.

stephentoub commented on src/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.Unix.cs:
```diff
@@ -0,0 +1,39 @@
+                {
+                    char ch = path[i];
+                    if (ch == '\\' || ch == '/' || ch == ':')
+                        return path.Substring(i + 1, length - i - 1);
```
This could be simplified to just:
```
return path.Substring(i + 1);
```
right? (Same question on the Substring in the Windows file.)

ianhays replied:
Yes, it can be.. I see no point in specifying the length if we're just going to be calculating it to the end of the string anyways.

I took this code from the [GetFileName method in System.Runtime.Extensions](https://github.com/dotnet/corefx/blob/master/src/System.Runtime.Extensions/src/System/IO/Path.cs)

stephentoub commented:
A few nits, otherwise LGTM.

bartonjs commented on src/System.IO.Compression/src/System/IO/Compression/ZipVersion.cs:
```diff
@@ -0,0 +1,24 @@
+
+    /// <summary>
+    /// The upper byte of the "version made by" flag in the central directory header of a zip file represents the 
+    /// OS of the system on which the zip was created. There are up 21 total OS values but we only use the three
```
Where do these 21 values come from?  If there's a RFC/ISO/etc, please link (if public) or identify (if behind a paywall).

ianhays replied:
The header tags for Windows and Unix (0,3) are equivalent to those used in ZLib and GZip so I got them from there. Beyond those values I'm not sure - the zip specification doesn't have an RFC/ISO that I could find/access.

Now that I've removed ```Undefined``` as a value, I'm going to rework this comment.

ianhays commented:
Thanks for the review, Steve and Jeremy. I've pushed an update addressing your feedback and rebased over master. Still need to squash.

Looks like we've got the desktop configurations included in the Open csprojs's now, so that raises the question of whether we want to include this change for 'net46' as well. It would be a breaking change. I've went ahead and included it in my last commit (since it's easier), but I can special-case for desktop if preferred.

ianhays commented:
Options for Desktop:
- Clone all of the old versions of these modified files to make them desktop specific. 
  - Lots of duplicate code.
- Add compiler IFDEFs around all the changes included to check if running on desktop. 
  - Messy.
- Use all these changes, but add a "ZipArchiveEntry.Desktop" with a ParseFileName that ignores the OS version tag of the Zip. 
  - Adds a bit of dead complexity
- Use all of these changes. 
  - Breaking change.

@weshaggard

stephentoub commented:
> Options for Desktop

Can you remind me what the "Desktop" build is?  My understanding was that the code for the full .NET Framework is forked... is that not the case, such that the full .NET Framework implementation of System.IO.Compression is sharing this code in corefx?

That'll impact my answer.  For example, if it's not actually the full .NET Framework build we're talking about, but rather some rarely used build that can be made available for certain uses, then "Use all of these changes" may very well be the right answer... but it's likely not the right answer if this actually is the full framework we're talking about.   In which case having a ZipArchiveEntry.Desktop.cs (or whatever the right name is) that provides an appropriately compatible implementation of ParseFileName and whatnot seems reasonable.

ianhays commented:
>such that the full .NET Framework implementation of System.IO.Compression is sharing this code in corefx?

That is my impression, yes. The System.IO.Compression.Desktop.csproj in TFS/FXCore/src/System.IO.Compression references the open source code for ZipArchive and Crc32Helper. It's why we had to add that FEATURE_ZLIB compiler directive. 

Aside: that Desktop csproj logic looks to have been copied into the open since there are now a bunch of "if framework == net46" checks present in the Open Sytem.IO.Compression.csproj.

stephentoub commented:
> The System.IO.Compression.Desktop.csproj

Yes, but DeflateStream and friends are in System.dll in the full framework, not in a separate System.IO.Compression.dll.  Hence my question as to what this other "desktop" library is.

ianhays commented:
@weshaggard @ericstj any insights/ideas?

weshaggard commented on src/System.IO.Compression/src/System.IO.Compression.csproj:
```diff
@@ -1,71 +1,92 @@
-  </ItemGroup>
-  <Import Project="$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory), dir.targets))\dir.targets" />
-</Project>
+    <PropertyGroup>
```
Looks like you converted the tabs to 4 spaces for all other projects and xml files in our repo it is 2 spaces can you please undo this?

ianhays replied:
Yes.

weshaggard commented:
@ericstj correct if I'm wrong but we are shipping the 4.1.0.0 version of this library for .NET Framework as an OOB library. It is a partial facade and so it does type-forward DeflateStream and GZipStream but the zip APIs are all implemented in this library.

ericstj commented:
@weshaggard yes it is OOB for desktop.  You still might consider an AppContext switch if you think someone might be broken by it (/cc @AlexGhiondea).

ianhays commented:
>You still might consider an AppContext switch if you think someone might be broken by it 

This PR and https://github.com/dotnet/corefx/issues/4991 are breaking changes to how we handle the paths within a Zip created on another system. 

Since we don't want to break, I'm going to special-case for back-compat by creating a ZipArchiveEntry.Desktop.cs included only in the "net46" configuration that keeps the old behavior for path handling.

weshaggard commented:
>ZipArchiveEntry.Desktop.cs

From a naming convention if you create that you should name it ZipArchiveEntry.net46.cs

ianhays commented:
>From a naming convention if you create that you should name it ZipArchiveEntry.net46.cs

ianhays commented:
Rebased to one commit and ready to merge if no one has any objections :)

ericstj commented:
Breaking changes can be just as much of a problem between platforms as they are on the same platform.  Folks want to write code that works everywhere.  If this is the right fix for desktop why not just do it and have an opt-out if someone gets broken?  Since the desktop implementation is in the package its not as breaking as pushing this out to a billion machines.

ianhays commented:
>why not just do it and have an opt-out if someone gets broken

I could certainly do that instead; I'd absolutely prefer consistency where possible but I'm concerned about two things on the AppContext switches:
- If I add the opt-out for net46 then won't I need to add it for the WindowsCore build as well for the "write code everywhere" aspect to be true? That means I'll have to add a dependency on System.AppContext for System.IO.Compression.
- There is an added overhead in doing an AppContext check every time we parse a file name. And we can't safely cache the value retrieved since it is subject to change during run time (via SetSwtich).

ericstj commented:
You don't need to opt out net46 since the library is oob and folks are already opting into the nupkg.

Caching is fine and is the reccomendation.  SetSwitch is only meant to be called by app in main.

AlexGhiondea commented:
We actually cache the value of the switch pretty much everywhere. 

As Eric pointed out, you should only call SetSwitch in the Main method.

Let me know if you need help setting up the AppContext switch.

ianhays commented:
>Caching is fine and is the reccomendation. SetSwitch is only meant to be called by app in main.

Ah okay that lessens my concerns for it then. That will be fine.

>You don't need to opt out net46 since the library is oob and folks are already opting into the nupkg.

To clarify, there are three places where we have ZipArchive code:
- Desktop: ZipAchive fork in fx/src/system.io.compression
- 4.1.0.0 OOB package: Is built by the "net46" configuration in Open csproj.
- CoreFX: Is built by the Open csproj when "net46"!="true".

I had not planned on pushing this change to Desktop but if I do then that will be a separate commit that requires no modification of Open code. I do, however, plan on applying this change to both configurations for which ZipArchive is built from Open code (i.e. net46 and non-net46). Those are the configurations that I was questioning whether special casing for back-compat is required, and from your recent comment it appears that it is not.

A separate but closely related issue is whether this fix should be applied to the desktop fork of ZipArchive under fx/src/System.IO.Compression. That issue will not require any AppContext switches to be added in Open code (i.e. this PR), but will require an AppContext switch be added to the forked code that is updated. That inclusion will modify the ZipArchiveEntry ParseFileName method to look like so:
```
internal static string ParseFileName(string path, ZipVersionMadeByPlatform madeByPlatform)
{
    // Allow users to always parse paths in a Zip based on the current OS instead of the target OS
    bool validateZipPathsBasedOnSourceOS;
    if (AppContext.TryGetSwitch("Switch.System.IO.Compression.ValidateZipPathsBasedOnSourceOS", out validateZipPathsBasedOnSourceOS) && !validateZipPathsBasedOnSourceOS)
    {
        return Path.GetFileName(path);
    }

    // Validation checking is done based on current OS, not source OS.
    PathInternal.CheckInvalidPathChars(path);
    
    if (madeByPlatform == ZipVersionMadeByPlatform.Unix)
    {
        int length = path.Length;
        for (int i = length; --i >= 0;)
            if (path[i] == '/')
                return path.Substring(i + 1);
        return path;
    }
    else
    {
        return Path.GetFileName(path);
    }
}
```
